File: code/SEACellsBenchmark/age/infer_stage_weights.py
# infer_stage_weights.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Dict, Iterable, List, Optional, Sequence, Tuple, Union


Number = Union[int, float]


@dataclass(frozen=True)
class StageWeights:
    """Return object for inferred stage weights."""
    stage_to_weight: Dict[str, float]
    ordered_stages: List[str]
    unit: str
    scale01: bool
    origin: str


def _natural_sort_key(s: str) -> Tuple:
    """
    Natural-ish sort key that keeps numeric chunks ordered numerically.
    Example: ["E6.5", "E10", "E8"] -> E6.5, E8, E10
    """
    parts = re.split(r"(\d+(?:\.\d+)?)", str(s))
    key: List[Tuple[int, Union[str, float]]] = []
    for p in parts:
        if p == "":
            continue
        try:
            key.append((1, float(p)))
        except ValueError:
            key.append((0, p.lower()))
    return tuple(key)


def _parse_embryo_day(stage: str) -> Optional[float]:
    """
    Parse embryo day from common labels:
      - "E6.5", "e8", "E10.0"
      - ranges: "E8.0-E8.5", "E8.0 to E8.5"  -> mean(8.0, 8.5)
      - bare numeric: "6.5" -> 6.5
    Returns None if cannot parse.
    """
    if stage is None:
        return None
    x = str(stage).strip().lower()

    # Capture E-values anywhere in the string, including ranges.
    # Examples matched: "e6", "e6.5", "e 8.0", "E8.0-E8.5"
    e_nums = re.findall(r"e\s*(\d+(?:\.\d+)?)", x)
    if e_nums:
        vals = [float(v) for v in e_nums]
        return sum(vals) / len(vals)

    # Fallback: if the whole stage looks numeric already
    if re.fullmatch(r"\d+(?:\.\d+)?", x):
        return float(x)

    return None


def infer_stage_weights(
    stages: Sequence[str],
    unit: str = "hours",
    scale01: bool = False,
    origin: str = "min",
    explicit_order: Optional[Sequence[str]] = None,
) -> StageWeights:
    """
    Infer a numeric weight w_s for each unique stage label.

    Parameters
    ----------
    stages:
        Iterable/sequence of stage labels from adata.obs['stage'] (strings).
    unit:
        One of {"hours", "days", "rank"}.
        - "hours": interpret stage as embryo day (E6.5 etc.) and convert to hours.
        - "days": interpret as embryo day and keep in days.
        - "rank": ignore numeric times; assign 0..m in chronological order of labels.
    scale01:
        If True, min-max scale weights to [0, 1] after origin/unit handling.
    origin:
        One of {"min", "absolute"}.
        - "min": subtract the minimum (earliest) value so the first stage is 0.
        - "absolute": keep absolute parsed values (e.g., E6.5 -> 6.5 days or 156 hours).
        (Note: for rank mode, origin is irrelevant.)
    explicit_order:
        Optional list giving the desired chronological order of stage labels.
        If provided, this order is used and parsing/sorting is skipped.

    Returns
    -------
    StageWeights:
        stage_to_weight: dict mapping stage label -> inferred w_s
        ordered_stages: list of stages in chronological order used for assignment
        unit, scale01, origin: echoes of settings

    Notes
    -----
    Recommended for developmental datasets:
      unit="hours", origin="min"  -> hours since earliest stage
    For categorical stages without parseable times:
      unit="rank"
    """
    unit = unit.lower()
    if unit not in {"hours", "days", "rank"}:
        raise ValueError("unit must be one of {'hours', 'days', 'rank'}")
    origin = origin.lower()
    if origin not in {"min", "absolute"}:
        raise ValueError("origin must be one of {'min', 'absolute'}")

    # Collect unique non-null labels
    uniq = []
    seen = set()
    for s in stages:
        if s is None:
            continue
        ss = str(s)
        if ss.strip() == "":
            continue
        if ss not in seen:
            uniq.append(ss)
            seen.add(ss)

    if len(uniq) < 2:
        raise ValueError("Need >= 2 unique non-empty stages to infer weights.")

    # Determine stage order
    if explicit_order is not None:
        ordered = [str(x) for x in explicit_order]
        missing = [x for x in uniq if x not in set(ordered)]
        if missing:
            raise ValueError(
                f"explicit_order is missing stage labels present in data: {missing}"
            )
    else:
        ordered = sorted(uniq, key=_natural_sort_key)

    # Try time-based weights if requested
    parsed_days: Dict[str, Optional[float]] = {s: _parse_embryo_day(s) for s in ordered}
    any_missing = any(parsed_days[s] is None for s in ordered)

    if unit == "rank" or any_missing:
        # Rank-based assignment: 0,1,2,... following `ordered`
        weights = {s: float(i) for i, s in enumerate(ordered)}
        used_unit = "rank" if unit == "rank" else "rank(fallback)"
    else:
        vals = [float(parsed_days[s]) for s in ordered]  # type: ignore[arg-type]
        if origin == "min":
            v0 = min(vals)
            vals = [v - v0 for v in vals]

        if unit == "hours":
            vals = [v * 24.0 for v in vals]
        # unit == "days" uses vals as-is

        weights = {s: float(v) for s, v in zip(ordered, vals)}
        used_unit = unit

    # Optional scaling to [0, 1]
    if scale01:
        ws = list(weights.values())
        wmin, wmax = min(ws), max(ws)
        if wmax == wmin:
            weights = {s: 0.0 for s in ordered}
        else:
            weights = {s: (weights[s] - wmin) / (wmax - wmin) for s in ordered}

    return StageWeights(
        stage_to_weight=weights,
        ordered_stages=ordered,
        unit=used_unit,
        scale01=scale01,
        origin=origin,
    )


if __name__ == "__main__":
    # Minimal example
    demo = ["E6.0", "E6.5", "E8.0-E8.5", "E10", "E6.0", None]
    sw = infer_stage_weights(demo, unit="hours", origin="min", scale01=False)
    print("Ordered:", sw.ordered_stages)
    print("Unit:", sw.unit)
    print("Weights:")
    for k in sw.ordered_stages:
        print(f"  {k:12s} -> {sw.stage_to_weight[k]:.2f}")

File: code/SEACellsBenchmark/age/infer_stage_weights.R
## =========================
## Packages
## =========================
suppressPackageStartupMessages({
  library(zellkonverter)         # readH5AD()
  library(SingleCellExperiment)  # colData()
  library(dplyr)
  library(tibble)
  library(stringr)
  library(gtools)                # mixedorder()
  library(ggplot2)
  library(readr)
})

## =========================
## Input
## =========================
# Option A (most common): read from an .h5ad file
h5ad_file <- "path/to/adata.h5ad"
sce <- zellkonverter::readH5AD(h5ad_file)

# Option B: if you already have a python AnnData object `adata` via reticulate:
# sce <- zellkonverter::AnnData2SCE(adata)

## =========================
## Helper: infer stage weights w_s
## =========================
infer_stage_weights <- function(stages,
                                unit = c("hours", "days", "rank"),
                                scale01 = FALSE,
                                origin = c("min", "absolute")) {
  unit   <- match.arg(unit)
  origin <- match.arg(origin)

  lev <- unique(as.character(stages))
  lev <- lev[!is.na(lev)]
  if (length(lev) < 2) stop("Need >= 2 unique stages to compute weights.")

  # Order stage labels in a human/numeric-friendly way (e.g., E6.5 < E10)
  lev <- lev[gtools::mixedorder(lev)]

  # Try to parse embryo day from labels like "E6.5" or ranges like "E8.0-E8.5"
  parse_embryo_day <- function(x) {
    x0 <- tolower(trimws(x))

    # Catch things like E6.5 or E8.0-E8.5 (take mean if multiple E-values)
    if (str_detect(x0, "e\\s*\\d")) {
      nums <- str_extract_all(x0, "(?<=e)\\s*\\d+\\.?\\d*")[[1]]
      nums <- gsub("\\s+", "", nums)
      if (length(nums) >= 1) return(mean(as.numeric(nums)))
    }

    # Fallback: purely numeric stage already (e.g., "6.5")
    suppressWarnings(num <- as.numeric(x0))
    if (!is.na(num)) return(num)

    return(NA_real_)
  }

  t_day <- vapply(lev, parse_embryo_day, numeric(1))

  if (unit == "rank" || any(is.na(t_day))) {
    # Rank-based assignment: 0,1,2,... following lev order
    w <- seq_along(lev) - 1
  } else {
    # Time-based assignment using embryo days
    w <- t_day
    if (origin == "min") w <- w - min(w)

    if (unit == "hours") w <- w * 24
    # unit == "days" leaves it in days
  }

  if (scale01) {
    if (max(w) == min(w)) {
      w <- rep(0, length(w))
    } else {
      w <- (w - min(w)) / (max(w) - min(w))
    }
  }

  tibble(stage = lev, w = w)
}

## =========================
## Extract obs (colData) and sanity-check columns
## =========================
meta <- as.data.frame(colData(sce))
stopifnot(all(c("celltype_new", "stage") %in% colnames(meta)))

meta <- meta %>%
  mutate(
    celltype_new = as.character(celltype_new),
    stage        = as.character(stage)
  )

## =========================
## k_{i,s} counts and within-celltype proportions k_s
## =========================
df_counts <- meta %>%
  count(celltype_new, stage, name = "k") %>%
  group_by(celltype_new) %>%
  mutate(k_prop = k / sum(k)) %>%   # this is k_s in the age-score formula
  ungroup()

## =========================
## Choose w_s
## =========================
# Recommended default for development datasets with E-labels:
#   unit="hours" -> hours since earliest stage
# If you want 0..1 scaling for cross-dataset comparability, set scale01=TRUE
stage_map <- infer_stage_weights(df_counts$stage,
                                 unit   = "hours",
                                 scale01 = FALSE,
                                 origin = "min")

df_counts <- df_counts %>%
  left_join(stage_map, by = "stage") %>%
  mutate(stage = factor(stage, levels = stage_map$stage))

## =========================
## Age_score per celltype: sum_s k_s * w_s
## =========================
df_age <- df_counts %>%
  group_by(celltype_new) %>%
  summarise(
    total_cells = sum(k),
    age_score   = sum(k_prop * w),
    .groups = "drop"
  ) %>%
  arrange(age_score)

## =========================
## Output (3): text table (celltype_new, age_score)
## =========================
age_table <- df_age %>% select(celltype_new, age_score)
print(age_table)
readr::write_tsv(age_table, "age_score_by_celltype.tsv")

## =========================
## Output (1): bar plot (counts), x=celltype_new, fill=stage
## =========================
p_bar <- ggplot(df_counts, aes(x = celltype_new, y = k, fill = stage)) +
  geom_col(width = 0.9) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(
    x = "celltype_new",
    y = "Cell population (k_{i,s} counts)",
    fill = "stage",
    title = "Cell population by cell type and stage"
  )

# Optional alternative: proportions instead of counts
p_bar_prop <- ggplot(df_counts, aes(x = celltype_new, y = k, fill = stage)) +
  geom_col(position = "fill", width = 0.9) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(
    x = "celltype_new",
    y = "Proportion within cell type",
    fill = "stage",
    title = "Stage composition within each cell type"
  )

## =========================
## Output (2): line plot (age_score vs celltype_new)
## =========================
df_age <- df_age %>%
  mutate(celltype_new = factor(celltype_new, levels = celltype_new)) # keep sorted

p_line <- ggplot(df_age, aes(x = celltype_new, y = age_score, group = 1)) +
  geom_line() +
  geom_point() +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(
    x = "celltype_new",
    y = "age_score",
    title = "Age score by cell type"
  )

## =========================
## Save plots
## =========================
ggsave("cell_population_by_stage_barplot.pdf", p_bar, width = 11, height = 5)
ggsave("cell_population_by_stage_barplot_prop.pdf", p_bar_prop, width = 11, height = 5)
ggsave("age_score_by_celltype_lineplot.pdf", p_line, width = 11, height = 4)

# If running interactively (RStudio), you can also just print plots:
print(p_bar)
print(p_bar_prop)
print(p_line)
